src/common_coin.rs

```diff
@@ -126,6 +126,9 @@ where
     }
 
     fn get_coin(&mut self) -> Result<()> {
+        if !self.netinfo.is_peer() {
```

Comment:
1. I'd vote against a flag: set membership checks should be fast, and a flag would be duplicating that information. I don't expect this to be a bottleneck.
2. I agree that it's a bit confusing at the moment but I don't know how a flag would help (`if self.is_peer` isn't that much more concise than `if self.netinfo.is_peer()`). I find the code in `agreement` kind of okay: `is_peer()` is checked exactly where we'd send a message, and prevents it for observers. I think I could arrange it in a similar way in `broadcast`. In `honey_badger`, the decryption share creation could be extracted into a separate method; that would turn the `is_peer()` branch a more readable one-liner. I don't have a good idea how the use in `common_coin` could be improved; there it's already only one occurrence, and the branch is one line.

Reply:
If the value of `self.netinfo.is_peer()` doesn't change during the lifetime of an algorithm, you could at least memoize the result of the membership check inside `NetworkInfo`, say, at construction time. I'm still not convinced that running the same computation in `is_peer` over and over again is a good idea. It's true that the computation is fast but it's not constant-time.

If you mean returning early in `broadcast` in case of observer nodes then yes, it might improve readability. Not all functions have the advantage of being split into two halves - one for both observer nodes and peer nodes, and one for peers only. In that case having an `is_peer` branch seems to be a better idea compared to returning `Ok(())` early for observer nodes.

Reply:
OK, I added the `is_peer` field and tried to simplify the branches a bit. Please have another look.